enrichments/trace/internal/elastic/span.go

```diff
@@ -0,0 +1,296 @@
+	span.Attributes().Range(func(k string, v pcommon.Value) bool {
+		switch k {
+		case semconv.AttributePeerService:
+			s.peerService = v.Str()
+		case semconv.AttributeMessagingDestinationName:
+			s.isMessaging = true
+			s.messagingDestn = v.Str()
+		case semconv.AttributeMessagingOperation:
+			s.isMessaging = true
+		case semconv.AttributeMessagingSystem:
+			s.isMessaging = true
+			s.messagingSystem = v.Str()
+		case semconv.AttributeMessagingDestinationTemporary:
+			s.isMessaging = true
+			s.messagingDestnTemp = true
+		case semconv.AttributeHTTPStatusCode,
+			semconv.AttributeHTTPResponseStatusCode:
+			s.isHTTP = true
+			s.httpStatusCode = v.Int()
+		case semconv.AttributeHTTPMethod,
+			semconv.AttributeHTTPRequestMethod,
+			semconv.AttributeHTTPURL,
+			semconv.AttributeHTTPTarget,
+			semconv.AttributeHTTPScheme,
+			semconv.AttributeHTTPFlavor,
+			semconv.AttributeNetHostName:
+			s.isHTTP = true
+		// TODO (lahsivjar): Is url ONLY present for HTTP? This is what is done in apm-data.
+		case semconv.AttributeURLFull:
+			s.isHTTP = true
+			// ignoring error as if parse fails then we don't want the url anyway
+			s.urlFull, _ = url.Parse(v.Str())
+		case semconv.AttributeURLScheme:
+			s.isHTTP = true
+			s.urlScheme = v.Str()
+		case semconv.AttributeURLDomain:
+			s.isHTTP = true
+			s.urlDomain = v.Str()
+		case semconv.AttributeURLPort:
+			s.isHTTP = true
+			s.urlPort = v.Int()
+		case semconv.AttributeURLPath:
+			s.isHTTP = true
+			s.urlPath = v.Str()
+		case semconv.AttributeURLQuery:
+			s.isHTTP = true
+			s.urlQuery = v.Str()
+		case semconv.AttributeRPCGRPCStatusCode:
+			s.isRPC = true
+			s.grpcStatus = codes.Code(v.Int()).String()
+		case semconv.AttributeRPCSystem:
+			s.isRPC = true
+			s.rpcSystem = v.Str()
+		case semconv.AttributeRPCService:
+			s.isRPC = true
+			s.rpcService = v.Str()
+		case semconv.AttributeDBStatement,
+			semconv.AttributeDBUser:
+			s.isDB = true
+		case semconv.AttributeDBName:
+			s.isDB = true
+			s.dbName = v.Str()
+		case semconv.AttributeDBSystem:
+			s.isDB = true
+			s.dbType = v.Str()
+		}
+		return true
+	})
+
+	// Ensure all dependent attributes are handled.
+	s.normalizeAttributes()
+
+	// Enrich the span depending on the nature of the span.
+	if isElasticTransaction(span) {
+		s.setTxnType(span)
+		s.setTxnResult(span)
+		s.setEventOutcome(span)
+	} else {
+		s.setEventOutcome(span)
+		s.setServiceTarget(span)
+	}
+}
+
+// normalizeAttributes sets any dependent attributes that
+// might not have been explicitly set as an attribute.
+func (s *Span) normalizeAttributes() {
+	if s.rpcSystem == "" && s.grpcStatus != "" {
+		s.rpcSystem = "grpc"
+	}
+}
+
+func (s *Span) setTxnType(span ptrace.Span) {
+	txnType := "unknown"
+	switch {
+	case s.isMessaging:
+		txnType = "messaging"
+		span.Attributes().PutStr("transaction.type", "messaging")
+	case s.isRPC, s.isHTTP:
+		txnType = "request"
+		// TODO (lahsivjar): DB type is not handled in apm-data
```

Comment:
> There is no such things as db transaction type in Elastic APM. The [transaction represents the entry](https://github.com/elastic/apm/blob/main/specs/agents/tracing-transactions.md) to a service - that works out nicely for incoming HTTP calls and e.g. for reading out from a message queue. But e.g. a database call always happens within a transaction - that can't be the entry point into a service - something else always needs to trigger the db call - that's how the Elastic APM data model work.

Thanks for the information here. My first thoughts were that if we instrumented a DB as a service then the transaction type would be `db` but now I realize that is not correct and in such cases, the txn would be w.r.t. the db as a service.

> So... here is my question: do we maybe want to try to solve this problem?

I don't yet grasp the whole issue but I think it would be worth to solve it if we can. I will go over the apm-data issue and create one here if required. We can continue the discussion there.

Reply:
If we're going to do this, does it belong here? I think maybe it would belong to the same processor, but adding artificial spans to the trace seems fairly different to what we're doing in this code.